File: sdk/js/src/modules/encoder.ts
/**
 * GRAFT Encoder Module — Typed JS wrapper
 */

import type { GraftDevice } from '../device.js'

export class GraftEncoder {
  private _device: ReturnType<GraftDevice['module']>

  constructor(device: GraftDevice) {
    this._device = device.module('encoder')
  }

  /** Attach encoder to two pins. */
  async attach(pinA: number, pinB: number): Promise<void> {
    await this._device.cmd(0x01, new Uint8Array([pinA, pinB]))
  }

  /** Read current count. */
  async read(): Promise<number> {
    const resp = await this._device.cmd(0x02)
    if (resp.length >= 4) {
      const view = new DataView(resp.buffer, resp.byteOffset, resp.byteLength)
      return view.getInt32(0, true)
    }
    return 0
  }

  /** Reset count to zero. */
  async reset(): Promise<void> {
    await this._device.cmd(0x03)
  }

  /** Listen for tick events. */
  onTick(handler: (count: number, delta: number) => void): void {
    this._device.onEvent((code, data) => {
      if (code === 0x01 && data.length >= 6) {
        const view = new DataView(data.buffer, data.byteOffset, data.byteLength)
        handler(view.getInt32(0, true), view.getInt16(4, true))
      }
    })
  }
}

File: sdk/python/src/graft/modules/encoder.py
"""GRAFT Encoder Module"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


class GraftEncoder:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def attach(self, pin_a: int, pin_b: int) -> None:
        payload = bytes([self._mod_id, 0x01, pin_a, pin_b])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def read(self) -> int:
        payload = bytes([self._mod_id, 0x02])
        resp = await self._device._send_command(CMD.MOD_CMD, payload)
        if len(resp) >= 5:  # module_id(1) + int32(4)
            return struct.unpack_from("<i", resp, 1)[0]
        return 0

    async def reset(self) -> None:
        payload = bytes([self._mod_id, 0x03])
        await self._device._send_command(CMD.MOD_CMD, payload)
